openhtf/core/test_state.py

```diff
@@ -130,7 +146,53 @@ def test_api(self):
                 running_phase_state.attachments,
                 running_phase_state.attach,
                 running_phase_state.attach_from_file,
-                self.notify_update))
+                self.get_measurement,
+                self.get_attachment,
+                self.notify_update,))
+
+  def get_measurement(self, measurement_name):
+    """Get a copy of a measurement value from current or previous phase.
+
+    Args:
+      measurement_name: str of the measurement name
+    Returns:
+      an ImmutableMeasurement
+    """
+    # The framework ignores measurements from SKIP and REPEAT phases
+    ignore_outcomes = set([test_record.PhaseOutcome.SKIP,])
+
+    # Check current running phase state
+    if self.running_phase_state:
+      if measurement_name in self.running_phase_state.measurements:
+        return ImmutableMeasurement.FromMeasurement(
+            self.running_phase_state.measurements[measurement_name])
+
+    # Iterate through phases in reversed order for LIFO.
+    for phase_record in reversed(self.test_record.phases):
+      if (phase_record.result not in ignore_outcomes and
+          measurement_name in phase_record.measurements):
+        measurement = phase_record.measurements[measurement_name]
+        return ImmutableMeasurement.FromMeasurement(measurement)
+
+    test.logger.warning('Could not find measurement: %s', measurement_name)
+    return None
+
+  def get_attachment(self, attachment_name):
+    """Get a copy of an attachment from current or previous phase."""
```

Comment:
is get_attachment() necessary, there's already access to a .attachments attribute?   if it's necessary, can you add some better documentation here? and it should function more closely to get_measurement(), ie, it should iterate over phases in reverse order, not the opposite order of get_measurement().

Reply:
From documentation on TestApi.measurements: "Only attachments that have been attached in the current phase show up here, and this attribute should not be modified directly;"

Added better documentation on this method.

No need to iterate through in reversed order, because attachment names are unique.  I did swap order so the running_phase_state it checked first (as it is in get_measurement).  I think of get_measurement as the special case and clarified comment to explain reversed.

Also swapped order of get_attachment and get_measurement.